engines/python/setup/djl_python/transformers_neuronx_scheduler/optimum_modeling.py

```diff
@@ -0,0 +1,284 @@
+#!/usr/bin/env python
+#
+# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file
+# except in compliance with the License. A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS"
+# BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or implied. See the License for
+# the specific language governing permissions and limitations under the License.
+# The below code is heavily inspired from Optimum Neuron under the following link:
+# https://github.com/huggingface/optimum-neuron/blob/v0.0.16-release/optimum/neuron/modeling.py
+
+import copy
+import torch
+from typing import Dict, Optional, Union
+from transformers import AutoModelForCausalLM, AutoConfig
+from transformers.utils import ModelOutput
+from transformers import GenerationConfig
+from transformers.generation import GenerationMixin
+from pathlib import Path
+from tempfile import TemporaryDirectory
+from transformers import PretrainedConfig
+from optimum.neuron.generation import TokenSelector
+from optimum.neuron.utils.version_utils import check_compiler_compatibility, get_neuronxcc_version
+from optimum.modeling_base import OptimizedModel
+
+
+class OptimumModelForCausalLM(OptimizedModel, GenerationMixin):
+    """
+        Overwrite of the NeuronModelForCausalLM to match legacy interface while bringing in new model generation methods.
+        Base class to convert and run pre-trained transformers decoder models on Neuron devices.
+
+        It implements the methods to convert a pre-trained transformers decoder model into a Neuron transformer model by:
+        - transferring the checkpoint weights of the original into an optimized neuron graph,
+        - compiling the resulting graph using the Neuron compiler.
+
+        Common attributes:
+            - model (`torch.nn.Module`) -- The decoder model with a graph optimized for neuron devices.
+            - config ([`~transformers.PretrainedConfig`]) -- The configuration of the original model.
+            - generation_config ([`~transformers.GenerationConfig`]) -- The generation configuration used by default when calling `generate()`.
+        """
+    auto_model_class = AutoModelForCausalLM
+    main_input_name = "input_ids"
+    CHECKPOINT_DIR = "checkpoint"
+    COMPILED_DIR = "compiled"
+
+    def __init__(self,
+                 model: torch.nn.Module,
+                 config: "PretrainedConfig",
+                 model_path: Union[str, "Path", "TemporaryDirectory"],
+                 generation_config: Optional[GenerationConfig] = None,
+                 **kwargs) -> None:
+        super().__init__(model, config)
+        self.model_path = model_path
+        self.cur_len = 0
+        self.checkpoint_dir = None
+        self.compiled_dir = None
```

Comment:
Where are these checkpoint_dir and compiled_dir used?

Reply:
Good catch - they are not now that the optimummodelforcausallm doesn't handle saving. I will remove those two attributes.